core-bundle/src/EventListener/DataContainer/TemplateOptionsListener.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener\DataContainer;
+
+use Contao\Controller;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\DataContainer;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+class TemplateOptionsListener
+{
+    /**
+     * @var Controller
+     */
+    private $controller;
+
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @var array
+     */
+    private $customTemplates;
+
+    /**
+     * @var string
+     */
+    private $templatePrefix;
+
+    /**
+     * @var string
+     */
+    private $proxyClass;
+
+    public function __construct(ContaoFramework $framework, RequestStack $requestStack, array $customTemplates, string $templatePrefix, string $proxyClass = null)
+    {
+        /** @var Controller $controller */
+        $controller = $framework->getAdapter(Controller::class);
+
+        $this->controller = $controller;
+        $this->requestStack = $requestStack;
+        $this->customTemplates = $customTemplates;
+        $this->templatePrefix = $templatePrefix;
+        $this->proxyClass = $proxyClass;
+    }
+
+    public function __invoke(DataContainer $dc): array
+    {
+        if ($this->isOverrideAll()) {
+            // Add a blank option that allows us to reset all custom templates to the default one
+            return array_merge(['' => '-'], $this->controller->getTemplateGroup($this->templatePrefix));
+        }
+
+        $defaultTemplate = $this->customTemplates[$dc->activeRecord->type] ?? $this->getLegacyDefaultTemplate($dc);
+
+        if (empty($defaultTemplate)) {
+            $defaultTemplate = $this->templatePrefix.$dc->activeRecord->type;
+        }
+
+        return $this->controller->getTemplateGroup($defaultTemplate.'_', [], $defaultTemplate);
+    }
+
+    /**
+     * Uses the reflection API to return the default template from a legacy class.
+     */
+    private function getLegacyDefaultTemplate(DataContainer $dc): ?string
+    {
+        if (null === $this->proxyClass || !method_exists($this->proxyClass, 'findClass')) {
+            return null;
+        }
+
+        $class = $this->proxyClass::findClass($dc->activeRecord->type);
+
+        if (empty($class) || $class === $this->proxyClass) {
+            return null;
+        }
+
+        $properties = (new \ReflectionClass($class))->getDefaultProperties();
+
+        return $properties['strTemplate'] ?? null;
```

Comment:
Not sure if we want to handle the case that someone might set the variable to a boolean for example?

Reply:
Via annotation it would not work because that would not validate. Not sure about validation of service tags otherwise.

Reply:
That would solve the potential problem.

Reply:
Booleans were never supported, so why account for this edge case now? It only makes the code more complicated. 🤷🏻‍♂️

Reply:
Agreed. The variable's type is specifically defined as string - so if some custom code or extension did in fact set this to something else, it needs to be fixed there. I'll revert it to the simpler code.